### PR TITLE
Introduce unified entrypoint for CLI scripts

### PR DESCRIPTION
**Pull request by rjernst**
CLI scripts have a common infrastructure in that they call to the shared
elasticsearch-cli shell script which launches them with the appropriate
java command line. However, each underlying Java class must implement
its own main method.

This commit introduces a single main method to be shared by CLIs. The
new CliToolLauncher takes in system properties to determine which tool
is being run, and a new CliToolProvider SPI allows defining and finding
the named tools.

relates #85758

### CODE REVIEW

**Comment by rjernst** [comment]
Note: this was spun off from #85758 to decrease the size of that PR. Most of the changes here are mechanical, just boilerplate adding the provider classes.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/docs 
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

**Comment by rjernst** [comment]
Some more suggestions for making reviewing easier. Sorry for the total number of lines.

I think all of the Provider subclasses and META-INF/services files can be skipped, they are just boilerplate. The shell script changes were mechanical, so they should be easy to skim through. Changes to the individual Command subclasses were also mechanical: remove the main method and make the class package protected.

The interesting files are really CliToolLauncher and CliToolProvider, which are very small.

Hope that helps.

**Comment by rjernst** [comment]
Some more suggestions for making reviewing easier. Sorry for the total number of lines.

I think all of the Provider subclasses and META-INF/services files can be skipped, they are just boilerplate. The shell script changes were mechanical, so they should be easy to skim through. Changes to the individual Command subclasses were also mechanical: remove the main method and make the class package protected.

The interesting files are really CliToolLauncher and CliToolProvider, which are very small.

Hope that helps.

File: x-pack/plugin/security/src/main/bin/elasticsearch-reconfigure-node.bat [hunk 1]
```diff
@@ -8,9 +8,8 @@ rem 2.0.
 setlocal enabledelayedexpansion
 setlocal enableextensions
 
-set ES_MAIN_CLASS=org.elasticsearch.xpack.security.cli.AutoConfigureNode
-set ES_ADDITIONAL_SOURCES=x-pack-env;x-pack-security-env
-set ES_ADDITIONAL_CLASSPATH_DIRECTORIES=lib/tools/security-cli
+set CLI_SCRIPT=%~0
+set CLI_LIBS=modules/x-pack-core,modules/x-pack-security,lib/tools/security-cli
 call "%~dp0elasticsearch-cli.bat" "--reconfigure" ^
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/bin/elasticsearch-reconfigure-node.bat:13]: The `reconfigure` option should not be passed here, right?

**Review by albertzaharovits — APPROVED** [review]
Besides https://github.com/elastic/elasticsearch/pull/85821/files#diff-4533746782f7e57dfcc0f2b37dfc394b525383628477bb4ba7a980452fc05317 
LGTM wrt to the `AutoConfigureNode` changes. Indeed only the `elasticsearch-reconfigure-node` cli was supposed to pass in the `--reconfigure` option.

**Review by ChrisHegarty — APPROVED** [review]
LGTM.

**Pull request merged**